tests/certification/configuration/redis/redis_test.go

```diff
@@ -0,0 +1,354 @@
+	"github.com/dapr/components-contrib/tests/certification/embedded"
+	"github.com/dapr/components-contrib/tests/certification/flow"
+	"github.com/dapr/components-contrib/tests/certification/flow/dockercompose"
+	"github.com/dapr/components-contrib/tests/certification/flow/network"
+	"github.com/dapr/components-contrib/tests/certification/flow/retry"
+	"github.com/dapr/components-contrib/tests/certification/flow/sidecar"
+	cu_redis "github.com/dapr/components-contrib/tests/utils/configupdater/redis"
+	configuration_loader "github.com/dapr/dapr/pkg/components/configuration"
+	"github.com/dapr/dapr/pkg/runtime"
+	dapr_testing "github.com/dapr/dapr/pkg/testing"
+	dapr "github.com/dapr/go-sdk/client"
+	"github.com/dapr/kit/logger"
+	redis "github.com/go-redis/redis/v8"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/dapr/components-contrib/tests/certification/flow/watcher"
+)
+
+const (
+	dockerComposeYAML             = "docker-compose.yml"
+	storeName                     = "configstore"
+	key1                          = "key1"
+	key2                          = "key2"
+	subscribeWaitDuration         = 3 * time.Second
+	subscribedMessageWaitDuration = 1000 * time.Millisecond
+	sidecarName1                  = "dapr-1"
+)
+
+// Cast go-sdk ConfigurationItem to contrib ConfigurationItem
+func castConfigurationItems(items map[string]*dapr.ConfigurationItem) map[string]*configuration.Item {
+	configItems := make(map[string]*configuration.Item)
+	for key, item := range items {
+		configItems[key] = &configuration.Item{
+			Value:    item.Value,
+			Version:  item.Version,
+			Metadata: item.Metadata,
+		}
+	}
+	return configItems
+}
+
+// Create UpdateEvent struct from given key, val pair
+func getUpdateEvent(key string, val string) configuration.UpdateEvent {
+	expectedUpdateEvent := configuration.UpdateEvent{
+		Items: map[string]*configuration.Item{
+			key: {
+				Value: val,
+			},
+		},
+	}
+	return expectedUpdateEvent
+}
+
+// Run redis commands and and add them in watcher
+func runRedisCommands(ctx flow.Context, updater *cu_redis.ConfigUpdater, messages *watcher.Watcher, basicTest bool) error {
+	scenarios := []struct {
+		cmd          []interface{}
+		want         [][]string
+		waitDuration time.Duration
+	}{
+		{
+			cmd:  []interface{}{"set", key1, "val1"},
+			want: [][]string{{key1, "val1"}},
+		},
+		{
+			cmd:  []interface{}{"mset", key1, "val1", key2, "val2"},
+			want: [][]string{{key1, "val1"}, {key2, "val2"}},
+		},
+		{
+			cmd:  []interface{}{"copy", key1, key2, "REPLACE"},
+			want: [][]string{{key2, "val1"}},
+		},
+		{
+			cmd:  []interface{}{"append", key1, "-append"},
+			want: [][]string{{key1, "val1-append"}},
+		},
+		{
+			cmd:  []interface{}{"setrange", key1, 4, "-offset"},
+			want: [][]string{{key1, "val1-offset"}},
+		},
+		{
+			cmd:          []interface{}{"expire", key1, 3},
+			want:         [][]string{{key1, "val1-offset"}, {key1, ""}},
+			waitDuration: 5 * time.Second,
```

Comment:
Q: Trying to understand the significance of waitDuration in test: Is it to keep updating values and to verify that subscription is happening correctly? 
And, here we are trying to test long wait duration particularly in this line?

Reply:
The general waitDuration is required to wait till the update to be received by subscriber. This is because we do a `Get` Call in subscribe handler (https://github.com/dapr/components-contrib/blob/dd06fcac9588b476d68e5064296e7743cd646d76/configuration/redis/redis.go#L337). If we execute another update (e.g. another `set` command) before this Get call executes, the value in the redis instance would change and so subscriber wouldn't receive the old value. So we wait after executing after every redis command to allow the handler to execute. 


In this specific line though, the reason is different. `expire` command generates two events: an `expire` event when the command executes and and `expired` event when the key actually expires. Here we have set the expire time to 3 seconds, so waiting for 5 seconds to collect both events.

Reply:
Thanks for explaining, got it.
Can we document this `expire` reasoning here. It may help in future to understand its significance.

Reply:
Done